### ISSUE TITLE
TypeLibrary: Add System.Type.MakeGenericType equivalent

### ISSUE THREAD

**Issue by peter-r-g** [Facepunch/sbox-issues#2508]
### What it is?

Currently, [Type.MakeGenericType](https://learn.microsoft.com/en-us/dotnet/api/system.type.makegenerictype?view=net-6.0) is not whitelisted and there is no `TypeLibrary` equivalent.

### What should it be?

There should be a `MakeGenericType` method within `TypeDescription`s that just points to the regular [Type.MakeGenericType](https://learn.microsoft.com/en-us/dotnet/api/system.type.makegenerictype?view=net-6.0) method. My use case is for custom networking and transmitting types that contain generics will need to be re-constructed on the other side using the method mentioned.

**Comment by garrynewman** [comment]
I don't think this is such a good idea from a whitelist point of view. It'd allow you to create any type based on a Type - which is really what the whole TypeLibrary system is trying to avoid.

**Comment by peter-r-g** [comment]
You would only be able to create types out of a base that is capable of containing generics. If security is a concern I'm a little confused because there is currently a way to create a generic instance of an object based on arbitrary `System.Type`s. If there's a way to slot this into the TypeLibrary using TypeDescriptions then that would be fine.

**Comment by garrynewman** [comment]
Yeah there could be a way to create an instance of the type, but I wouldn't particularly want you to be able to do to that from the type itself.

**Comment by peter-r-g** [comment]
That isn't what `Type.MakeGenericType` is though. It creates a new `System.Type` out of a base generic type with an array of its generic components. For example:

```cs
var baseType = typeof( Dictionary<,> );
var components = new Type[] {
    typeof( string),
    typeof( int )
}

var type = baseType.MakeGenericType( components );
Log.Info( type);
// System.Collections.Generic.Dictionary`2<System.String, System.Int32>
```

As I mentioned previously if it has to be implemented through `TypeDescription`s that's fine. I just need at least some way of getting info of a created type and creating an instance of said type.
